Specify allowed characters for Baggage keys and values (#3801)
open-telemetry/opentelemetry-specification: specification/baggage/api.md

user-defined properties. Each name in `Baggage` MUST be associated with
exactly one value.

> Baggage **names** are strings that satisfy the `token` definition from

Reviewer: The W3C is currently considering allowing UTF-8 strings for keys as well as values. The otel API can make their own decision, but I think the UTF-8 change is likely to go through in W3C at this point.

Author: Please link a ticket. Such change in W3C will be seriously breaking.

Author: https://github.com/w3c/baggage/issues/99

There is no resolution there as of Jan 2024. I don't think it should block this PR - it makes changes to match the __current__ W3C spec. If we adopt ahead of time more extensive alphabet for keys, then OTEL simply won't be compatible w/ W3C spec.

Reviewer 2: @dyladan Would the change be breaking? 

As right now a key like: `%20` is valid and will not be encoded. I guess after https://github.com/w3c/baggage/issues/99 it would be still valid but the key will be encoded to a whitespace: ` `.

Reviewer: Im sorry I didn't follow up. We're working with the i11y group at w3c but it seems this change will in fact not go through in the baggage spec. It is not providing enough value to justify breaking our existing implementations.

Reviewer 2: I think this conversation can be resolved.

[RFC7230, Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).

> Alpha-numeric names are strongly recommended. Language API MAY return

Reviewer 2: nit

```suggestion
Alpha-numeric names are strongly recommended. Language API SHOULD return
```

Specification, § 3.2.1.1](https://www.w3.org/TR/baggage/#baggage-string)
which allows duplicate entries for a given name.

Baggage **names** are strings that satisfy the `token` definition from
[RFC7230, Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).

> Alpha-numeric names are strongly recommended. Language API SHOULD return

Author: SHOULD is a normative word that does not apply to user-supplied values, it's not something that the OTEL SDK can control.

an error when given a baggage name that is not a `token`.

Baggage **values** are any valid UTF-8 strings. Language API MUST accept

> any valid UTF-8 string as baggage value. Language API MUST guarantee that

Author: phrased it differently

> `Get` and `Set` operations are symmetrical with respect to name and value,

Reviewer 3: ```suggestion
```

Author: phrased it differently

> i.e. `Get` MUST return the same **value** string as received in `Set`.

Author: "i.e." is not used to give an example, but to provide alternative wording. But you're right, "symmetrical" was unnecessary, I phrased it in simpler terms, and provided explicit examples (thanks for a good idea!)

Specification, § 3.2.1.1](https://www.w3.org/TR/baggage/#baggage-string)
which allows duplicate entries for a given name.

Baggage **names** are strings that satisfy the `token` definition from
[RFC7230, Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).
Alpha-numeric names are strongly recommended. Language API SHOULD return

> an error when given a baggage name that is not a `token`.

Reviewer 4: > Java is also not compliant with this. Currently, we only propagate keys/values that meet this definition, but we allow arbitrary keys for in-memory usage.

➕ 

Our stance in java is that the API should never return an error. If anyone uses the API incorrectly, we log a warning and use a no-op implementation of the operation. Applying this logic to baggage and these new key constraints, when a invalid key / value is recorded in baggage, we would log an error and skip recording that key / value pair. Like @jkwatson said, this is not our behavior today and skipping recording those key / value pairs is a breaking behavioral change. But because we only propagate the valid key / values, the consequence of us not enforcing these constraints is only that other consumers of baggage in process (i.e. span processors, log processors) can access key / values that will not be propagated.

@jkwatson I propose that if this is merged, we keep the existing behavior which records the invalid key / values (since changing would be breaking) but add a check which logs if the key / value is invalid. If invalid, log a message indicating that the key / value won't be propagated, but record it for access by in-process consumers.

Author: I am fine changing SHOULD to MAY

Reviewer 5: Actually, API methods in OpenTelemetry MUST NOT return errors. A first principle of OpenTelemetry API design is that users should never have to wrap any OpenTelemetry API call in error or exception handling code.

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles

Author: > API methods in OpenTelemetry MUST NOT return errors

they must not raise exceptions, but they _can_ return errors when idiomatic in the language (e.g. Try in Scala)

Author: Proposed change:

> Language API MUST ignore baggage entries given a name that is not a `token`, e.g. they MAY return an error (if idiomatic for the language) or fallback to a no-op implementation.

Reviewer 4: > Applying this logic to baggage and these new key constraints, when a invalid key / value is recorded in baggage, we would log an error and skip recording that key / value pair. Like @jkwatson said, this is not our behavior today and skipping recording those key / value pairs is a breaking behavioral change.

Sorry if my message wasn't clear. Invalid key / values aren't propagated across application boundaries, but they are recorded in baggage such that they can be used by other baggage consumers like processors. 

I believe your intent here would be to say that not only should invalid key / value pairs not be propagated, but they also shouldn't be recorded in baggage in the first place. The "not recorded in baggage in the first place" part is what I think is a breaking change for java.

Author: I see. The question of how encoding restrictions impact in-process API is always a tricky one, especially since encodings technically could vary across propagators. But we want to ensure that at least w3c propagator works, hence the `token` restriction. The main function of baggage is not in-process propagation, but across processes - for in-process people should use Context API directly with no restrictions. The fact that Java Baggage implementation provided accidental feature of unrestricted in-process propagation should not be treated as a breaking change if that behavior is restricted, imo, as long as it still maintains its cross-process propagation capability (the primary function).

Reviewer 2: Should the Baggage API have the same constrains as W3C Baggage Propagator? Is this change not mixing different concerns?

The Baggage API can be used in-process and then usage of UTF-8 keys would not be a problem.

> The main function of baggage is not in-process propagation, but across processes - for in-process people should use Context API directly with no restrictions. 

I will say that is it "one" of the functions. Even the [spec](https://opentelemetry.io/docs/concepts/signals/baggage/) says:

> Baggage MAY be propagated across process boundaries or across any arbitrary boundaries

And [here](https://opentelemetry.io/docs/concepts/signals/baggage/#why-does-otel-baggage-exist):

> Baggage provides a uniform way to store and propagate information across a trace and other signals. For example, you may want to attach information from your application to a span and retrieve that information much later and use it later on with another span.

Shouldn't the W3C Baggage Propagator report a problem (e.g. emit some logs) when it encounters a key that it cannot encode?

Author: Changed to allow any strings.

> an error when given a baggage name that is not a `token`.

Reviewer 2: Should the Baggage API have the same constrains as W3C Baggage Propagator? Is this change not mixing different concerns?

The Baggage API can be used intra-process and then usage of UTF-8 keys would not be a problem.

Shouldn't the W3C Baggage Propagator report a problem (e.g. emit some logs) when it encounters a key that it cannot encode?

Specification, § 3.2.1.1](https://www.w3.org/TR/baggage/#baggage-string)
which allows duplicate entries for a given name.

Baggage **names** are any valid UTF-8 strings. Language API SHOULD NOT
restrict which strings are used as baggage **names**. However, the
specific `Propagator`s that are used to transmit baggage entries across
component boundaries may impose their own restrictions on baggage names.

> For example, the [W3C Baggage specification](https://github.com/w3c/baggage/blob/main/baggage/HTTP_HEADER_FORMAT.md#key)

Reviewer 2: When possible I think we should refer to [W3C Working Draft](https://www.w3.org/standards/types#WD)

```suggestion
For example, the [W3C Baggage specification](https://www.w3.org/TR/baggage/#key)
```

restricts the baggage keys to strings that satisfy the `token` definition
from [RFC7230, Section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).
For maximum compatibility, alpha-numeric names are strongly recommended
to be used as baggage names.

Baggage **values** are any valid UTF-8 strings. Language API MUST accept
any valid UTF-8 string as baggage **value** in `Set` and return the same
value from `Get`.

Language API MUST treat both baggage names and values as case sensitive.

> See also [W3C Baggage -- Case sensitivity of keys](https://github.com/w3c/baggage/blob/main/baggage/HTTP_HEADER_FORMAT_RATIONALE.md#case-sensitivity-of-keys).

Reviewer 2: I suggest to clearly say that this is from Rationale docs.

```suggestion
See also [W3C Baggage Rationale](https://github.com/w3c/baggage/blob/main/baggage/HTTP_HEADER_FORMAT_RATIONALE.md#case-sensitivity-of-keys).
```
